day15/src/main.rs

```diff
@@ -0,0 +1,275 @@
+use std::{fs, str::FromStr, usize};
+
+const INPUT_FILE_PATH: &str = "./input";
+
+#[derive(Debug)]
+struct Move {
+    repr: char,
+    x: i32,
+    y: i32,
+}
+
+#[derive(Clone)]
+struct Vec2 {
+    x: usize,
+    y: usize,
+}
+
+#[derive(Clone, PartialEq)]
+enum Cell {
+    PLAYER,
+    WALL,
+    BOX,
+    EMPTY,
+}
+
+impl ToString for Cell {
+    fn to_string(&self) -> String {
+        format!(
+            "{}",
+            match self {
+                Self::EMPTY => ".",
+                Self::BOX => "O",
+                Self::PLAYER => "@",
+                Self::WALL => "#",
+            }
+        )
+    }
+}
+
+impl FromStr for Cell {
+    type Err = ();
+
+    fn from_str(input: &str) -> Result<Cell, Self::Err> {
+        match input {
+            "#" => Ok(Cell::WALL),
+            "@" => Ok(Cell::PLAYER),
+            "O" => Ok(Cell::BOX),
+            "." => Ok(Cell::EMPTY),
+            _ => Err(()),
+        }
+    }
+}
+
+struct Board {
+    size: Vec2,
+    data: Vec<Cell>,
+    player_position: Vec2,
+}
+
+impl Board {
+    fn new(width: usize, height: usize) -> Board {
+        let mut data = Vec::with_capacity(width * height);
+        data.resize(width * height, Cell::EMPTY);
+
+        Board {
+            data,
+            size: Vec2 { x: width, y: height },
+            player_position: Vec2 { x: 0, y: 0 },
+        }
+    }
+
+    fn attempt_move_player(&mut self, player_move: Move) {
+        let new_pos = Board::add_vec_to_move(&self.player_position, &player_move);
+        let next_cell = self.get(new_pos.x, new_pos.y).unwrap();
+
+        match next_cell {
+            Cell::PLAYER => panic!("Something went terribly wrong"),
+            Cell::EMPTY => {
+                self.set(self.player_position.x, self.player_position.y, Cell::EMPTY);
+                self.set(new_pos.x, new_pos.y, Cell::PLAYER);
+                self.player_position = new_pos;
+
+                println!("Moved player")
+            }
+            Cell::BOX => {
+                let chain_length = self.get_movable_box_chain_length(&new_pos, &player_move, 0);
+
+                if chain_length > 0 {
+                    self.set(self.player_position.x, self.player_position.y, Cell::EMPTY);
+                    self.set(new_pos.x, new_pos.y, Cell::PLAYER);
+                    self.player_position = new_pos.clone();
+                }
+
+                let mut end_of_chain = new_pos.clone();
+
+                for _ in 0..chain_length {
+                    end_of_chain = Board::add_vec_to_move(&end_of_chain, &player_move);
+                }
+
+                self.set(end_of_chain.x, end_of_chain.y, Cell::BOX);
+
+                println!("Moving chain: {}", chain_length)
+            }
+            Cell::WALL => {
+                println!("Wall collision")
+            }
+        }
+    }
+
+    fn add_vec_to_move(vec: &Vec2, mv: &Move) -> Vec2 {
+        Vec2 {
+            x: (vec.x as i32 + mv.x) as usize,
```

Comment:
The casting makes me very sad

Reply:
And it highlights a bug: If `vec.x` is bigger than `i32::MAX` the calculation will be off.